README.md
# MySQL Sniffer [中文介绍](https://github.com/Qihoo360/mysql-sniffer/blob/master/README_CN.md)
MySQL Sniffer is a network traffic analyzer tool for MySQL, it is developed by Qihoo DBA and infrastructure team. This commandline tool captures and analyzes packets destined for a MySQL 
server or Client, and outputs them in a standard log format including access time, users, IP, database, query_time, rows number and query.

MySQL Sniffer also analyzer Atlas's network traffic. Atlas is a MySQL protocol-based database middleware project，github：[https://github.com/Qihoo360/Atlas](https://github.com/Qihoo360/Atlas)

# System requirements:
- Certified to run on CentOS v6 
- Commandline access to the server with root privileges

# Usage
```
./mysql-sniffer -h
Usage mysql-sniffer [-d] -i eth0 -p 3306,3307,3308 -l /var/log/mysql-sniffer/ -e stderr
         [-d] -i eth0 -r 3000-4000
         -d daemon mode.
         -s how often to split the log file(minute, eg. 1440). if less than 0, split log everyday
         -i interface. Default to eth0
         -p port, default to 3306. Multiple ports should be splited by ','. eg. 3306,3307
            this option has no effect when -f is set.
         -r port range, Don't use -r and -p at the same time
         -l query log DIRECTORY. Make sure that the directory is accessible. Default to stdout.
         -e error log FILENAME or 'stderr'. if set to /dev/null, runtime error will not be recorded
         -f filename. use pcap file instead capturing the network interface
         -w white list. dont capture the port. Multiple ports should be splited by ','.
         -t truncation length. truncate long query if it's longer than specified length. Less than 0 means no truncation
         -n keeping tcp stream count, if not set, default is 65536. if active tcp count is larger than the specified count, mysql-sniffer will remove the oldest one
```
# Install
```
git clone git@github.com:Qihoo360/mysql-sniffer.git
cd mysql-sniffer
mkdir proj
cd proj
cmake ../
make
cd bin/
```
# Dependence 
glib2-devel、libpcap-devel、libnet-devel

# Install

```
git clone git@github.com:Qihoo360/mysql-sniffer.git
cd mysql-sniffer
mkdir proj
cd proj
cmake ../
make
cd bin/
```

# More
More MySQL Sniffer information, Atlas and some other technology please pay attention to our Hulk platform official account or QQ:104180820 

<img src="http://i.imgur.com/pL4ni57.png" width = "500" alt="2">

# Thanks
Thanks for the contributions [yihaoDeng](https://github.com/yihaoDeng) and [winkyao](https://github.com/winkyao) have made for MySQL Sniffer


libnids-1.24/CHANGES
v1.24 Mar 14 2010
- fixed another remotely triggerable NULL dereference in ip_fragment.c
- unofficial patch that enables tracking of already established TCP connections
- missing reset of some tcp_* variables upon nids_exit
- correct calculation of radiotap header
- compilation warning fixes with newer gcc
- use pcap_get_selectable_fd() instead of pcap_fileno()

v1.23 Feb 23 2008
- fixed remotely triggerable NULL dereference in ip_fragment.c
- fix DLT_PRISM_HEADER linkoffset calculation
- check for DATA_FRAME_IS_QOS in wireless frames
- free queued tcp segments with too old seq

v1.22 Jul 22 2007
- in TCP stream, the byte with absolute offset 0 was treated as urgent data;
fixed
- DLT_IEEE802_11_RADIO handling
- added a few missing checks for failed malloc
 
v1.21 May 10 2006
- more externals to access libnids' intrinsics from the outside
- nids_unregister_*()
- UDP checksumming fix (0 is not an error according to RFC768)
- nids_params.tcp_workarounds
- nids_params.multiproc and queue_limit: merged a patch which creates a 
separate thread for packet capture; 
- in killtcp.c, send two more RST packets (required because of MS05-019
patch)
- glibc 2.4 syslog.h disaster workaround

v1.20 Feb 4 2005
- added wscale option parsing; surprisingly, it seems to be in some use
- added nids_dispatch(), for systems which do not ignore pcap timeout
- ability to specify hosts/networks for which we do not check checksums

v1.19 Aug 08 2004
- fixed signed/unsigned comparisons; 1.18 could be possibly crashed in tcp
  options parsing (though an unlikely to happen memory layout is required);
  now the source is compiled with -W -Wall
- export pcap header of the last received packet (to get timestamp etc)
- export the timeout parameter to pcap_open_live in params
- support DLT_PRISM_HEADER
- support DLT_PPP_SERIAL
- let through dataless acks
- fixed raw_init() prototype
- switched to use %edi instead of %ebx in csum_partial to make gcc-3.5 happy
  when compiling with -fPIC; cleaned inline asm
- fixed a bug when a queued FIN segment was not processed properly, which
  resulted in not closing a stream

v1.18 Oct 15 2003
- reject tcp packets with old timestamp; needed to pass fragroute test;
  well, linux 2.0.36 did not support this ;)
- fixed memory corruption which could be caused by overlarge TCP packets
- adjusted checksum.c to not use multiline literals (for gcc 3.3)
- in configure.in, even if found libnet files, try compilation; there is
  another library with the same name
- fix a bug in "collect" field handling; if you did collect-- and then
  collect++ (which is rare), you would get a single junk packet
- correct handling of exec_prefix in configure.in
- unlink config.status in "make distclean"
- use pcap_hdr->caplen instead pcap_hdr->len; the only gain seems to be to
  gracefully handle pcap files with too short snaplen
- changed soname to libnids.so.1.x, as binary compatibility is not
  guaranteed
- switched to sourceforge as homepage

v1.17 Dec 12 2002
- fixed a stupid bug in TCP reassembly; having received a particular order
  of TCP out of frame segments, libnids could lost track of the current 
  seq, and miss the following data stream
- DLT_FDDI
- benign typo in hash.c
- mentioned usefulness of two process buffering on a fast network

v1.17rc1 Aug 30 2002
- support for libnet-1.1 and --with-libnet=no
- added support for libpcap save files
- finally, DLT_LINUX_SLL is recognized
- removed a horrible assumption on sizeof(pointer); it could result in
  segfault in scan.c
- --enable-shared
- __i386 -> __i386__ || __i386 :(
- support for 802.1Q VLAN 
- support for wireless frames (DLT_IEEE802_11)
- got rid of (obsolete) pcap_open_live_new
- bail out if link type is unknown, instead of pretending it is ethernet
- $(MAKE) -> $(MAKE) $(AM_MAKEFLAGS)
- added a working link to Ptacek-Newsham paper
- %hi -> %hu :)
- align IP header if necessary (should not be)
- improved libraries detection
- mentioned usefulness od setsockopt(...SO_RCVBUF...) on a fast network

v1.16 Nov 3 2000
- nah, at least a release forced by a security bug. A typo in libnids.c
could cause libnids to segfault when source routed frame has been received.

v1.15 Oct 9 2000
- token ring support
- new configurable option (non-default): if a tcp callback hasn't processed 
  all available data, it is called immediately again
- fixed alignment in hash.c, which caused sigsegv on Sparc
- another _obviously_ redundant include file added to configure test progs
- html version of the API documentation

v1.14 Jun 28 2000
- fixed memory leak in tcp.c (queued tcp segments used to be not freed
  after connection termination)
- added support to capture packets on all interfaces, including loopback
  (linux only, using new libpcap features - autoconf changed)
- added nids_register_udp(); if anyone cares for UDP checksums...
- stupid bug in nids_register_ip_frag() fixed
- removed comments from asm code in checksum.c; Solaris compiler didn't
  recognized them - sigh
- signed/unsigned bug in scan.c fixed
- tcp callback could be notified even if no nw data arrived - fixed
- added ability to disable tcp processing
- added ability to refrain from setting promisc flag
- libc5 support
- alpha platform support
- now it's possible to do setuid(nobody) after nids_init() with no loss
  of functionality (killtcp works)
- removed pcap_lookupnet() call - one can capture packets from an interface
  with no IP assigned
- hash function in tcp.c with pseudorandom parameters
- #define NIDS_MAJOR 1, #define NIDS_MINOR 14 in nids.h

v1.13 Jan 18 2000
- Changes by Dug Song:
  - GNU autoconf support
  - code cleanup and new libnids(3) manpage
  - disable portscan detection if scan_num_hosts == 0
  - new field in nids_params for pcap(3) support: pcap_filter
  - subtle bugfix in ip_check_ext()
  - Solaris support (endianness fixes, etc.)
- another tiny check in tcp.c 

v1.12 Sep 15 1999
- processing of ICMP Destination Unreachable 

src/sniff-config.cpp
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <signal.h>
#include <errno.h>
#include <iostream>
#include <string>
#include <vector>

#include "sniff-config.h"
#include "util.h"

using std::string;
using std::vector;
using std::cout;
using std::endl;

typedef struct{
    string filter;
    string device;
    string port_range;
    string log_dir;
    string pcapfile;
    vector<int> ports;

    FILE* error_log;    
    int port_type;      /* range or enum */
    int split;          /* whether split the log */
    int daemon;         /* daemon mode */
    int error_log_on;
    int minute;          /* how often we want to change the date_flag */
    sig_atomic_t time_count;  /* check the flag to determine whether the date has changed */
    int truncate_len;
    int tcp_stream_cnt;
}sniff_config;

static sniff_config config;

void config_sigalarm_handler(int sig){
    config.time_count++;
    /* call me next time */
    alarm(config.minute * 60);
}

void split(string& s, const string& delim, vector< string >& ret) {  
    size_t last = 0;  
    size_t index = s.find_first_of(delim,last);  
    while (index != string::npos) {  
        ret.push_back(s.substr(last, index-last));  
        last = index + 1;  
        index = s.find_first_of(delim,last);  
    }  
    if (index-last > 0) {  
        ret.push_back(s.substr(last,index-last));  
    }  
}  

static int parse_cmdline_port_option(const char* optarg, const char* fragment){
    vector<string> ports;

    config.port_range.assign(optarg);
    split(config.port_range, ",", ports);
    if(ports.size() == 0){
        return -1;
    }

    char tmp[32];
    int port;

    config.filter = "";
    config.ports.clear();
    port = atoi(ports[0].c_str());
    if(port > 0 || port < 65536){
        snprintf(tmp, 32, "tcp port %d", port);
        config.ports.push_back(port);
        config.filter += tmp;
    }
    for(size_t i = 1;i < ports.size(); i++){
        port = atoi(ports[i].c_str());
        if(port <= 0 || port > 65535){
            continue;
        }
        config.ports.push_back(port);        
        snprintf(tmp, 32, " or port %d ", port);
        config.filter += tmp;
    }
    config.filter += string(" or ") + fragment;
    return 0;
}


static int parse_cmdline_port_range(const char* optarg, const char* fragment){
    vector<string> ports;
    config.port_range.assign(optarg);
    
    split(config.port_range, "-", ports);
    if(ports.size() == 2){ 
        config.ports[0] = atoi(ports[0].c_str());
        config.ports[1] = atoi(ports[1].c_str());
    }else{
        return -1;
    }
    
    if(config.ports[0] < 1 || config.ports[0] > 65535 
            || config.ports[1] < 1 || config.ports[1] > 65535){
        return -1;
    }
    if(config.ports[0] > config.ports[1]){
        int tmp = config.ports[0];
        config.ports[0] = config.ports[1];
        config.ports[1] = tmp;
    }

    config.port_type = PORT_RANGE;
    config.filter = "";
    config.filter += string("tcp portrange ") + optarg;
    return 0;
}

static int parse_cmdline_white_list(const char* optarg, string& filter){
    vector<string> ports;
    string whitelist(optarg);

    filter = "";
    split(whitelist, ",", ports);
    char tmp[32];
    for(size_t i = 0;i < ports.size();i++){
        snprintf(tmp, 32, " and not port %d ", atoi(ports[i].c_str()));
        filter += tmp;
    }
    return 0;
}

static void print_help(const char* argv0){
    fprintf(stderr,
            "Usage %s [-d] -i eth0 -p 3306,3307,3308 -l /var/log/mysql-sniffer/ -e stderr\n" 
            "         [-d] -i eth0 -r 3000-4000\n" 
            "         -d daemon mode.\n"
            "         -s how often to split the log file(minute, eg. 1440). if less than 0, split log everyday\n"
            "         -i interface. Default to eth0\n"
            "         -p port, default to 3306. Multiple ports should be splited by ','. eg. 3306,3307\n"
            "            this option has no effect when -f is set.\n"
            "         -r port range, Don't use -r and -p at the same time\n"
            "         -l query log DIRECTORY. Make sure that the directory is accessible. Default to stdout.\n"
            "         -e error log FILENAME or 'stderr'. if set to /dev/null, runtime error will not be recorded\n"
            "         -f filename. use pcap file instead capturing the network interface\n"
            "         -w white list. dont capture the port. Multiple ports should be splited by ','.\n"
            "         -t truncation length. truncate long query if it's longer than specified length. Less than 0 means no truncation\n"
            "         -n keeping tcp stream count, if not set, default is 65536. if active tcp count is larger than the specified count, mysql-sniffer will remove the oldest one\n",
            argv0);
}

static int is_logdir_accessible(const char* logdir){
    DIR* dir = opendir(logdir);
    if(dir){
        /* directory exists */
        closedir(dir);
        if(access(logdir, R_OK|W_OK|X_OK) < 0){
            return -1;
        }
        return 0;
    }else if(ENOENT == errno){
        /* directory does not exist */
        return -1;
    }else{
        /* opendir() failed */
        return -1;
    }
    /* never come here */
    return 0;
}

static int parse_cmdline_option(int argc, char** argv){
    int opt;
    int ret = 0;
    const char* filter_fragment = "(ip[6:2] & 0x1fff != 0)";
    
    int opt_len;
    string whitelist;
    while((opt = getopt(argc, argv, "dhi:s:p:l:e:f:r:w:t:n:")) != -1 && ret != -1){
        switch(opt){
            case 'd':
                config.daemon = 1;
                break;
            case 's':
                config.split = 1;
                config.minute = atoi(optarg);
                break;
            case 'i':
                config.device = optarg;
                break;
            case 'p':
                ret = parse_cmdline_port_option(optarg, filter_fragment);
                break;
            case 'l':
                config.log_dir = optarg;
                ret = is_logdir_accessible(optarg);
                if(ret == -1){
                    fprintf(stderr, "the log directory is not accessible!\n");
                }
                break;
            case 'e':
                opt_len =  strlen(optarg);
                if(!strcmp(optarg, "stderr")){
                    config.error_log = stderr;
                }else{
                    config.error_log = fopen(optarg, "wa+");
                    if(config.error_log == NULL){
                        ret = -1;
                    }
                }
                break;
            case 'f':
                config.pcapfile = optarg;
                break;
            case 'r':
                ret = parse_cmdline_port_range(optarg, filter_fragment);
                break;
            case 'w':
                ret = parse_cmdline_white_list(optarg, whitelist);
                break;
            case 't':
                config.truncate_len = atoi(optarg);
                break;
            case 'n':
                config.tcp_stream_cnt = atoi(optarg);
                break;
            case 'h':
            default:
                print_help(argv[0]);
                ret = -1;
                break;
        }
    }
    if(config.split == 1 && ret != -1){
        config_set_log_split();
    }
    if(config.truncate_len <=0){
        /* int max */
        config.truncate_len = 0x7fffffff;
    }
    config.filter += whitelist;

    return ret;
}


int config_init(int argc, char** argv){
    /* default option */
    config.split = 0;
    config.daemon = 0;
    config.error_log_on = 1;
    config.error_log = NULL;

    config.ports.push_back(3306);
    config.port_type = PORT_ENUM;

    config.device = "eth0";
    config.port_range = "3306";
    config.log_dir = "stdout";
    config.pcapfile = "";
    config.filter = "tcp port 3306 or (ip[6:2] & 0x1fff != 0)";

    config.truncate_len = 0;

    config.time_count = 0;
    config.minute = 24 * 60;
    config.tcp_stream_cnt = 65536;
    return parse_cmdline_option(argc, argv);
}

void config_print(){
    cout<<"device: "<<config.device<<endl;
    if(config.port_type == PORT_RANGE){
        printf("min port: %d, max port: %d\n", config.ports[0], config.ports[1]);
    }else{
        printf("ports: ");
        for(size_t i = 0;i < config.ports.size(); i++){
            printf("%d ", config.ports[i]);
        }
        printf("\n");
    }
    cout<<"filter: "<<config.filter<<endl;
    cout<<"log: "<<config.log_dir<<endl;
    cout<<"pcapfile: "<<config.pcapfile<<endl;
}


void config_fini(){
    if(config.error_log != NULL){
        fflush(config.error_log);
        fclose(config.error_log);
    }
}

void config_set_log_split(){
    if(config.minute <= 0){
        config.minute = 24 * 60;
    }
    signal(SIGALRM, config_sigalarm_handler);
    alarm(config.minute * 60);
}

int config_is_log_split(){
    return config.split;
}

int config_is_daemon(){
    return config.daemon;
}

FILE* config_get_err_log(){
    return config.error_log;
}

const char* config_get_device(){
    return config.device.c_str();
}
const char* config_get_pcapfile(){
    return config.pcapfile.c_str();
}

const char* config_get_logdir(){
    return config.log_dir.c_str();
}

const char* config_get_filter(){
    return config.filter.c_str();
}

sig_atomic_t config_get_time_count(){
    return config.time_count;
}

int config_get_port_type(){
    return config.port_type;
}

int config_is_server_port(int port){
    if(config.port_type == PORT_ENUM){
        for(size_t i = 0;i < config.ports.size(); i++){
            if(port == config.ports[i]){
                return 1;
            }
        }
    }else if(port >= config.ports[0] && port <= config.ports[1]){
        return 1;
    }
    return 0;
}

int config_get_truncate_len(){
    return config.truncate_len;
}

int config_get_tcp_stream_cnt() {
    return config.tcp_stream_cnt;
}
